lib/clientutils.js

```diff
@@ -57,60 +57,68 @@ var updateState = function(state, newState, callback) {
 
 // process state (following plugin execution)
 var processState = function(r, callback) {
-  if (r.abort) {
-    // => client has called for the request to be aborted
-    if (r.response) {
-      debug('Client issued abort during plugin execution.');
-      r.response.abort();
-    }
-    callback(new Error('Client issued abort')); // callback with error to prevent request retry
-  } else if (r.state.abortWithResponse) {
-    // => a plugin has requested an abort and an alternative response be returned to the client
+  var abort = function() {
     if (r.response) {
+      debug('Client issued abort.');
       r.response.abort();
     }
+  };
+
+  if (r.abort) {
+    // [1] => Client has called for the request to be aborted.
+    abort();
+    callback(new Error('Client issued abort')); // no retry
+    return;
+  }
+
+  if (r.state.abortWithResponse) {
+    // [2] => Plugin requested abort and specified alternative response.
+    abort();
     sendResponseToClient(r.state.abortWithResponse, r.clientStream, r.clientCallback);
     var err = new Error('Plugin issued abort');
     err.skipClientCallback = true;
-    callback(err); // callback with error to prevent request retry
-  } else if (r.state.retry && r.state.attempt < r.state.maxAttempt) {
-    // => one or more plugins have called for the request to be retried
-    if (r.response) {
-      r.response.abort();
-    }
+    callback(err); // no retry
+    return;
+  }
+
+  if (r.state.retry && r.state.attempt < r.state.maxAttempt) {
+    // [3] => One or more plugins have called for the request to be retried.
+    abort();
     debug('Plugin issued a retry.');
-    callback(); // callback without error in order to retry the request
-  } else {
-    // => the response is "good" and can be returned to the awaiting client
-    //    OR
-    //    too many retry attempts; the "bad" response must be returned to the client
-    if (r.response) {
-      r.clientStream.abort = function() {
-        debug('Client issued abort.');
-        r.response.abort(); // monkey-patch real abort
-      };
-    }
-    if (r.state.retry) {
-      debug('Too many retry attempts.');
-      r.state.retry = false;
-    }
-    if (!r.state.sending) {
-      // request has not been sent, implies we are processing 'onRequest' hooks
-      callback(); // callback without error to proceed with executing the client's request
-    } else {
-      // processing 'onError' or 'onResponse' hooks
-      if (r.eventRelay) {
-        // pass response through to awaiting client
-        r.eventRelay.resume();
-      }
-      if (r.clientStream.destinations.length > 0) {
-        // pipe response body to client stream
-        r.response.pipe(r.clientStream.passThroughReadable);
-      }
-      r.response.resume();
-      callback(new Error('No retry requested')); // no retry
-    }
+    callback(); // retry
+    return;
+  }
+
+  if (r.response) {
+     // monkey-patch real abort function
```

Comment:
Leading space, probably worth removing - I'm surprised the linter didn't complain.